File: katas/03-bank_ocr/gabriel.py
number_0 = " _ " \
           "| |" \
           "|_|"
number_1 = "   " \
           "  |" \
           "  |"
number_2 = " _ " \
           " _|" \
           "|_ "
number_3 = " _ " \
           " _|" \
           " _|"
number_4 = "   " \
           "|_|" \
           "  |"
number_5 = " _ " \
           "|_ " \
           " _|"
number_6 = " _ " \
           "|_ " \
           "|_|"
number_7 = " _ " \
           "  |" \
           "  |"
number_8 = " _ " \
           "|_|" \
           "|_|"
number_9 = " _ " \
           "|_|" \
           " _|"

translator = {
    number_0: "0",
    number_1: "1",
    number_2: "2",
    number_3: "3",
    number_4: "4",
    number_5: "5",
    number_6: "6",
    number_7: "7",
    number_8: "8",
    number_9: "9"
}


def validate_account_number(account_number):
    if "?" in account_number:
        return False
    checksum = 0
    for index, char in enumerate(reversed(list(account_number))):
        checksum += (index+1) * int(char)
    return checksum % 11 == 0


def get_numbers(file):
    numbers = {0: '', 1: '', 2: '', 3: '', 4: '', 5: '', 6: '', 7: '', 8: ''}
    for line in file:
        index = 0
        for i in range(0, 27, 3):
            if (not line[i:i + 3] and line != "\n") or line == "\n":
                numbers[index] += "   "
            else:
                numbers[index] += line[i:i + 3].replace("\n", " ")
            index += 1
    return numbers


def translate_numbers(numbers):
    account_number = ''
    for number in numbers:
        try:
            account_number += translator[numbers[number]]
        except KeyError:
            account_number += "?"
    return account_number


def read_file(file_path):
    f = open(file_path)
    numbers = get_numbers(f)
    f.close()
    return translate_numbers(numbers)


def get_status(file_path):
    account_number = read_file(file_path)
    if "?" in account_number:
        return account_number + " ILL"
    if not validate_account_number(account_number):
        return account_number + " ERR"
    return account_number


def do_rotation(number):
    list_possible = []
    for index, char in enumerate(number):
        possible = list(number)
        for value in [" ", "_", "|"]:
            possible[index] = value
            if "".join(possible) in translator:
                list_possible.append(translator["".join(possible)])
    return list(dict.fromkeys(list_possible))


def get_values(numbers, index):
    account_number = translate_numbers(numbers)
    possible_values = []
    for option in do_rotation(numbers[index]):
        possible = list(account_number)
        possible[index] = option
        if validate_account_number("".join(possible)):
            possible_values.append("".join(possible))
    return possible_values


def get_possibles(file_path):
    f = open(file_path)
    numbers = get_numbers(f)
    f.close()
    account_number = translate_numbers(numbers)
    possible_values = []
    if "?" in account_number:
        for index in [i for i, e in enumerate(list(account_number)) if e == "?"]:
            possible_values += get_values(numbers, index)
    elif not validate_account_number(account_number):
        for index in range(len(account_number)):
            possible_values += get_values(numbers, index)
    if not possible_values:
        return account_number + " ILL"
    elif len(possible_values) == 1:
        return possible_values[0]
    else:
        return account_number + " AMB " + str(sorted(possible_values))

File: katas/03-bank_ocr/gabriel.js
const readline = require('readline');
const fs = require('fs');

const number_0 = " _ " +
                 "| |" +
                 "|_|";
const number_1 = "   " +
                 "  |" +
                 "  |";
const number_2 = " _ " +
                 " _|" +
                 "|_ ";
const number_3 = " _ " +
                 " _|" +
                 " _|";
const number_4 = "   " +
                 "|_|" +
                 "  |";
const number_5 = " _ " +
                 "|_ " +
                 " _|";
const number_6 = " _ " +
                 "|_ " +
                 "|_|";
const number_7 = " _ " +
                 "  |" +
                 "  |";
const number_8 = " _ " +
                 "|_|" +
                 "|_|";
const number_9 = " _ " +
                 "|_|" +
                 " _|";

const translator = {};
translator[number_0] = "0";
translator[number_1] = "1";
translator[number_2] = "2";
translator[number_3] = "3";
translator[number_4] = "4";
translator[number_5] = "5";
translator[number_6] = "6";
translator[number_7] = "7";
translator[number_8] = "8";
translator[number_9] = "9";

async function get_numbers(file){
    let numbers = {0: '', 1: '', 2: '', 3: '', 4: '', 5: '', 6: '', 7: '', 8: ''};
    let index = 0;
    await new Promise((resolve) => {
        file.on('line', async (line) => {
            index = 0;
            for (let i=0; i<27; i+=3){
                if (line.substring(i, i+3)){
                    numbers[index] += line.substring(i, i+3);
                    for (let j=0; j<(3-line.substring(i, i+3).length); j++){
                        numbers[index] += " ";
                    }
                } else {
                    numbers[index] += "   ";
                }
                index+=1;
            }
            resolve(numbers);
        });
    });
    return numbers
}

async function translate_numbers(numbers){
    let account_number = '';
    return new Promise((resolve)=>{
        for (let i=0; i<9; i++) {
            if (numbers[i] in translator){account_number += translator[numbers[i]]}
            else{account_number += '?'}
        }
        resolve(account_number);
    });
}


function do_rotation(number){
    let list_possible = [];
    for (let i=0; i<number.length; i++){
        let possible = number.split('');
        ["_", " ", "|"].forEach( (item) => {
            possible[i] = item;
            if (possible.join('') in translator){
                list_possible.push(translator[possible.join('')]);
            }
        });
    }
    return [...new Set(list_possible)]
}

async function get_values(numbers, index){
    let account_number = await translate_numbers(numbers);
    let possible_values = [];
    do_rotation(numbers[index]).forEach(async (option) => {
        let possible = account_number.split('');
        possible[index] = option;
        let valid_account_number = await validate_account_number(possible.join(''));
        if (valid_account_number){
            possible_values.push(possible.join(''))
        }
    });
    return possible_values
}

//User Story #1
async function read_file(file_path){
    const readInterface = await readline.createInterface({
        input: fs.createReadStream(file_path)
    });
    return await translate_numbers(await get_numbers(readInterface));
}

//User Story #2
async function validate_account_number(account_number){
    if (account_number.includes("?")){
        return false
    }
    let checksum = 0;
    for (let i=0; i<account_number.length; i++){
        checksum += Number(account_number[i]) * (9-i)
    }
    return checksum % 11 === 0
}

//User Story #3
async function get_status(file_path){
    let account_number = await read_file(file_path);
    if (account_number.includes('?')){return account_number + ' ILL'}
    else if (!validate_account_number(account_number)) {return account_number + ' ERR'}
    else {return account_number}
}

// User Story #4
async function get_possibles(file_path){
    const readInterface = await readline.createInterface({
        input: fs.createReadStream(file_path)
    });
    let numbers = await get_numbers(readInterface);
    let account_number = await translate_numbers(numbers);
    let possible_values = [];
    return new Promise(async (resolve) => {
        let valid_account_number = await validate_account_number(account_number);
        if (!valid_account_number || account_number.includes('?')){
            for (let i=0; i<account_number.length; i++){
                let new_values = await get_values(numbers, i);
                possible_values = possible_values.concat(new_values);
            }
        }
        if (!possible_values){
            resolve(account_number + "ILL")
        } else if (possible_values.length === 1){
            resolve(possible_values[0])
        } else {
            resolve(account_number + " AMB ['" +
                possible_values.sort().join("', '") + "']")
        }
    });
}

module.exports = {
    read_file,
    validate_account_number,
    get_status,
    get_possibles
};
